src/graphql-types/ReasonApolloQuery.re

```diff
@@ -1,84 +1,105 @@
-module type InternalConfig = {let apolloClient: ApolloClient.generatedApolloClient;};
+[@bs.module] external gql : ReasonApolloTypes.gql = "graphql-tag";
 
-module QueryFactory = (InternalConfig:InternalConfig) => {
-    external castResponse : string => {. "data": Js.Json.t } = "%identity";            
-    external asJsObject : 'a => Js.t({..}) = "%identity";
+[@bs.module]
+external shallowEqual : (Js.Json.t, Js.Json.t) => bool =
+  "fbjs/lib/shallowEqual";
 
-    [@bs.module] external gql : ReasonApolloTypes.gql = "graphql-tag";
-    [@bs.module] external shallowEqual : (Js.t({..}), Js.t({..})) => bool = "fbjs/lib/shallowEqual";
+module type InternalConfig = {
+  let apolloClient: ApolloClient.generatedApolloClient;
+};
 
-    
-    type response =
-      | Loading
-      | Loaded(Js.Json.t)
-      | Failed(string);
-
-    type state = { 
-      response: response, 
-      variables: Js.Json.t
+module QueryFactory = (InternalConfig: InternalConfig) => {
+  exception InvalidState;
+  type response =
+    | Loading
+    | Loaded(Js.Json.t)
+    | Failed(string);
+  type state = {
+    response,
+    queryObservable: ref(ApolloClient.watchQueryObservable),
+    subscription: ref(option(ApolloClient.subscription))
+  };
+  type action =
+    | UpdateResponse(response);
+  type retainedProps = {variables: Js.Json.t};
+  let component =
+    ReasonReact.reducerComponentWithRetainedProps("ReasonApollo");
+  let make = (~query as q, children) => {
+    let convertResponse = response =>
+      switch (
+        Js.to_bool(response##loading),
+        Js.Nullable.to_opt(response##data),
+        Js.Nullable.to_opt(response##error)
+      ) {
+      | (true, _, None) => Loading
+      | (false, _, Some(error)) => Failed(error)
```

Comment:
Can loading be true while having an error? Maybe `(_, _, Some(error))` is more safe... (didn't checked react-apollo for this part)